handler/src/main/java/io/netty/handler/ssl/util/BouncyCastlePemReader.java

```diff
@@ -0,0 +1,163 @@
+import java.security.PrivilegedAction;
+import java.security.Provider;
+
+import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
+import org.bouncycastle.openssl.PEMDecryptorProvider;
+import org.bouncycastle.openssl.PEMEncryptedKeyPair;
+import org.bouncycastle.openssl.PEMException;
+import org.bouncycastle.openssl.PEMKeyPair;
+import org.bouncycastle.openssl.PEMParser;
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
+import org.bouncycastle.openssl.jcajce.JceOpenSSLPKCS8DecryptorProviderBuilder;
+import org.bouncycastle.openssl.jcajce.JcePEMDecryptorProviderBuilder;
+import org.bouncycastle.operator.InputDecryptorProvider;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.pkcs.PKCS8EncryptedPrivateKeyInfo;
+import org.bouncycastle.pkcs.PKCSException;
+
+public final class BouncyCastlePemReader {
+
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(BouncyCastlePemReader.class);
+    private static Provider bcProvider;
+    private static Boolean available;
+
+    static {
+        tryLoading();
+    }
+
+    public static boolean isAvailable() {
+        return available;
+    }
+
+    private static void tryLoading() {
+        available = AccessController.doPrivileged(new PrivilegedAction<Boolean>() {
+            @Override
+            public Boolean run() {
+                try {
+                    Class<Provider> bcProviderClass
+                      = (Class<Provider>) Class.forName("org.bouncycastle.jce.provider.BouncyCastleProvider",
+                              true, this.getClass().getClassLoader());
+                    bcProvider = bcProviderClass.newInstance();
+                    logger.debug("Bouncy Castle provider available");
+                    return Boolean.TRUE;
+                } catch (Exception e) {
+                    logger.debug("Cannot load Bouncy Castle provider", e);
+                    return Boolean.FALSE;
+                }
+            }
+        });
+    }
+
+    public static PrivateKey getPrivateKey(InputStream keyInputStream, String keyPassword) {
+        try {
+            return getPrivateKey(newParser(keyInputStream), keyPassword);
+        } catch (Exception e) {
+            logger.debug("Unable to extract private key", e);
+            return null;
+        }
+    }
+
+    public static PrivateKey getPrivateKey(File keyFile, String keyPassword) {
+        try {
+            return getPrivateKey(newParser(keyFile), keyPassword);
+        } catch (Exception e) {
+            logger.debug("Unable to extract private key", e);
+            return null;
+        }
+    }
+
+    private static PrivateKey getPrivateKey(PEMParser pemParser, String keyPassword) throws PEMException, IOException,
+            PKCSException, OperatorCreationException {
+        try {
+            Object object = pemParser.readObject();
+            logger.debug("Parsed PEM object of type " + object.getClass().getName() + " and assume key is "
```

Comment:
is this debug statement necessary? if so can you use the slf4j logging format to avoid always doing string concatenation operations (e.g. `"Parsed PEM object of type {} ...", object.getClass().getName()`